walkeeper/src/timeline.rs

```diff
@@ -1,50 +1,84 @@
-//! This module contains tools for managing timelines.
-//!
+//! This module contains timeline id -> safekeeper state map with file-backed
+//! persistence and support for interaction between sending and receiving wal.
 
 use anyhow::{bail, Result};
 use fs2::FileExt;
 use lazy_static::lazy_static;
 use log::*;
-use postgres_ffi::xlog_utils::{find_end_of_wal, TimeLineID};
+use postgres_ffi::xlog_utils::find_end_of_wal;
 use std::cmp::{max, min};
 use std::collections::HashMap;
 use std::fs::{self, File, OpenOptions};
-use std::io::{Seek, SeekFrom};
-use std::path::Path;
+use std::io::{Seek, SeekFrom, Write};
 use std::sync::{Arc, Condvar, Mutex};
 use zenith_utils::bin_ser::LeSer;
 use zenith_utils::lsn::Lsn;
+
 use zenith_utils::zid::ZTimelineId;
 
-use crate::receive_wal::{SafeKeeperInfo, CONTROL_FILE_NAME, SK_FORMAT_VERSION, SK_MAGIC};
 use crate::replication::{HotStandbyFeedback, END_REPLICATION_MARKER};
+use crate::safekeeper::{
+    AcceptorProposerMessage, ProposerAcceptorMessage, SafeKeeper, SafeKeeperState, Storage,
+    SK_FORMAT_VERSION, SK_MAGIC,
+};
 use crate::WalAcceptorConf;
+use postgres_ffi::xlog_utils::{XLogFileName, XLOG_BLCKSZ};
+
+const CONTROL_FILE_NAME: &str = "safekeeper.control";
 
 /// Shared state associated with database instance (tenant)
 #[derive(Debug)]
 struct SharedState {
-    /// quorum commit LSN
+    /// Safekeeper object
+    sk: SafeKeeper<FileStorage>,
+    /// opened file control file handle (needed to hold exlusive file lock)
+    control_file: File,
+    /// For receiving-sending wal cooperation
+    /// quorum commit LSN we've notified walsenders about
     commit_lsn: Lsn,
-    /// information about this safekeeper
-    info: SafeKeeperInfo,
-    /// opened file control file handle (needed to hold exlusive file lock
-    control_file: Option<File>,
     /// combined hot standby feedback from all replicas
     hs_feedback: HotStandbyFeedback,
 }
 
 impl SharedState {
-    fn new() -> Self {
-        Self {
+    /// Restore SharedState from control file. Locks the control file along the
+    /// way to prevent running more than one instance of safekeeper on the same
+    /// data dir.
+    /// If create=false and file doesn't exist, bails out.
+    fn create_restore(
+        conf: &WalAcceptorConf,
+        timelineid: ZTimelineId,
+        create: bool,
```

Comment:
Rust doesn't have named arguments, which means that functions with ancillary bool arguments are inherently less ergonomic than those taking well-named enums. This is what user is going to see at the call site:

```rust
foo.create_restore(bar, baz, true);
```

Compare this to the following:

```rust
// Alternatively, enum RestoreOptions { CreateControlFile, NoCreateControlFile };
bitflags! {
  struct RestoreOptions: u32 {
    const CreateControlFile = 1;
  }
}

foo.create_restore(bar, baz, RestoreOptions::CreateControlFile | ...);
```

Reply:
Nice idea, did like that. Initially I was like 'that's better for sure', but the resulting diff is `33 insertions(+), 14 deletions(-)` made me a bit less enthusiastic about it though.